rust/chains/hyperlane-sealevel/src/mailbox.rs

```diff
@@ -0,0 +1,754 @@
+                }),
+                commitment: Some(CommitmentConfig::finalized()),
+                min_context_slot: None,
+            },
+            with_context: Some(false),
+        };
+        let accounts = self
+            .rpc_client
+            .get_program_accounts_with_config(&self.mailbox.program_id, config)
+            .await
+            .map_err(ChainCommunicationError::from_other)?;
+
+        // Now loop through matching accounts and find the one with a valid account pubkey
+        // that proves it's an actual message storage PDA.
+        let mut valid_message_storage_pda_pubkey = Option::<Pubkey>::None;
+
+        for (pubkey, account) in accounts.iter() {
+            let unique_message_pubkey = Pubkey::new(&account.data);
+            let (expected_pubkey, _bump) = Pubkey::try_find_program_address(
+                mailbox_dispatched_message_pda_seeds!(unique_message_pubkey),
+                &self.mailbox.program_id,
+            )
+            .ok_or_else(|| {
+                ChainCommunicationError::from_other_str(
+                    "Could not find program address for unique_message_pubkey",
+                )
+            })?;
+            if expected_pubkey == *pubkey {
+                valid_message_storage_pda_pubkey = Some(*pubkey);
+                break;
+            }
+        }
+
+        let valid_message_storage_pda_pubkey =
+            valid_message_storage_pda_pubkey.ok_or_else(|| {
+                ChainCommunicationError::from_other_str(
+                    "Could not find valid message storage PDA pubkey",
+                )
+            })?;
+
+        // Now that we have the valid message storage PDA pubkey, we can get the full account data.
+        let account = self
+            .rpc_client
+            .get_account_with_commitment(
+                &valid_message_storage_pda_pubkey,
+                CommitmentConfig::finalized(),
+            )
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .value
+            .ok_or_else(|| {
+                ChainCommunicationError::from_other_str("Could not find account data")
+            })?;
+        let dispatched_message_account =
+            DispatchedMessageAccount::fetch(&mut account.data.as_ref())
+                .map_err(ChainCommunicationError::from_other)?
+                .into_inner();
+        let hyperlane_message =
+            HyperlaneMessage::read_from(&mut &dispatched_message_account.encoded_message[..])?;
+
+        Ok((
+            hyperlane_message,
+            LogMeta {
+                address: self.mailbox.program_id.to_bytes().into(),
+                block_number: dispatched_message_account.slot,
+                // TODO real values?
+                // It's inconvenient to get these :|
+                block_hash: H256::zero(),
+                transaction_hash: H256::zero(),
+                transaction_index: 0,
+                log_index: U256::zero(),
```

Comment:
Also looks important to fix?

Reply:
Kicking can to https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2527